### PR TITLE
Bump minimum numpy version to 1.20

### LINKED ISSUE #13872: Bump minimum required Numpy to 1.20?
Following [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html), which we adopted in [APE 18](https://github.com/astropy/astropy-APEs/blob/main/APE18.rst), I think we can now require numpy>1.20? If so, it would be good to do this and clean up any Numpy<1.20-related code before the 5.2 feature freeze. I won't have time to do this but if anyone wants to champion it, please go ahead! :smile:

### PR DESCRIPTION
**Pull request by mhvk**
### Description

As the title suggests, set the minimum version for numpy to 1.20, following NEP29/APE18.
I did the removal for numpy 1.18 and 1.19 separately so that if necessary we can stick to >=1.19 (though I don't think we should do that).

I hope I got everything...

Fixes #13872

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by mhvk** [comment]
p.s. Actually not sure any more how this is dealt with in the changelog. Should there be an entry? If so, where?

**Comment by mhvk** [comment]
Hmm, strange error at https://github.com/astropy/astropy/actions/runs/3308804140/jobs/5461453938
Since this is with random data, I'm just going to rerun the test. Just in case the log gets lost, the error is below.

```
self = <astropy.modeling.tests.test_models.TestFittable1DModels object at 0x7f9cf537ea30>
model_class = <class 'astropy.modeling.powerlaws.BrokenPowerLaw1D'>
Name: BrokenPowerLaw1D
N_inputs: 1
N_outputs: 1
Fittable parameters: ('amplitude', 'x_break', 'alpha_1', 'alpha_2')
test_parameters = {'constraints': {'fixed': {'x_break': True}}, 'log_fit': True, 'parameters': [1, 1, 2, 3], 'x_lim': [0.1, 100], ...}
fitter = <astropy.modeling.fitting.TRFLSQFitter object at 0x7f9ce91633a0>

    @pytest.mark.skipif(not HAS_SCIPY, reason='requires scipy')
    @pytest.mark.parametrize('fitter', fitters)
    def test_fitter1D(self, model_class, test_parameters, fitter):
        """
        Test if the parametric model works with the fitter.
        """
        fitter = fitter()
    
        x_lim = test_parameters['x_lim']
        parameters = test_parameters['parameters']
        model = create_model(model_class, test_parameters)
    
        if isinstance(parameters, dict):
            parameters = [parameters[name] for name in model.param_names]
    
        if "log_fit" in test_parameters:
            if test_parameters['log_fit']:
                x = np.logspace(x_lim[0], x_lim[1], self.N)
        else:
            x = np.linspace(x_lim[0], x_lim[1], self.N)
    
        np.random.seed(0)
        # add 10% noise to the amplitude
        relative_noise_amplitude = 0.01
        data = ((1 + relative_noise_amplitude * np.random.randn(len(x))) *
                model(x))
        new_model = fitter(model, x, data)
    
        # Only check parameters that were free in the fit
        params = [getattr(new_model, name) for name in new_model.param_names]
        fixed = [param.fixed for param in params]
        expected = np.array([val for val, fixed in zip(parameters, fixed)
                             if not fixed])
        fitted = np.array([param.value for param in params
                           if not param.fixed])
        if model_class == models.BrokenPowerLaw1D:
            atol = 3.75
        else:
            atol = self.fit_error
>       assert_allclose(fitted, expected, atol=atol)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=3.75
E       
E       Mismatched elements: 1 / 3 (33.3%)
E       Max absolute difference: 11.22491935
E       Max relative difference: 5.61245968
E        x: array([ 1.019002, 13.224919,  3.005807])
E        y: array([1, 2, 3])
```

**Comment by dhomeier** [comment]
`atol=3.75` seems a very crude fit to begin with.

**Comment by mhvk** [comment]
Ah, the seed is actually set so it fails again. I suspect this is a problem with a particular combination of scipy/numpy. Advise very welcome! cc @nden, @WilliamJamieson

**Comment by WilliamJamieson** [comment]
> astropy/astropy/actions/runs/3308804140/jobs/5461453938

The `BrokenPowerLaw1D` being flaky in this test strikes again. I'm not surprised that this is failing here again as it was a bit sensitive. If you look I already had to give it a "special" tolerance. The way the fitting problem is being posed here is bit ill posed for the `TRFLSQFitter` (the fixed jump maybe making things strange).

> Ah, the seed is actually set so it fails again. I suspect this is a problem with a particular combination of scipy/numpy.

I think you are right about the strange combination of scipy/numpy. Do you mind if I push a few commits to this PR to see what I can do for the failure? Its easier to use the CI then to test all the combinations locally in this case.

**Comment by mhvk** [comment]
@WilliamJamieson - yes, please go ahead and test here!

**Comment by nden** [comment]
@WilliamJamieson The modeling changes LGTM

**Comment by WilliamJamieson** [comment]
@mhvk I think the modeling changes are ready. I think the issue is using `scipy 1.5`. The TRF algorithm appears to have been improved between `1.5` and `1.6` which is likely why this worked in all the other CI tests.

**Comment by pllim** [comment]
But we're doing this so close to feature freeze. Should we hold off till v6 and do it earlier in the dev cycle?

Will this catch downstream packages off-guard?

**Comment by mhvk** [comment]
scipy 1.6.0 was released end of Dec 2020, so that will be 2 years ago by the time we release 5.2. Should we just bump the minimum version? Though I'm also fine with the work-around here.

@pllim - I'm not sure I'm following: what is wrong with narrowing down the numpy/scipy versions for 5.2? We're not changing anything here. The tests here just showed that some version of scipy in fact didn't quite work -- we just didn't know because it wasn't tested. So, I guess the real worry is whether we should exclude some scipy versions for stable...

**Comment by pllim** [comment]
We did not give people who test against dev astropy a lot of notice, dropping support for older versions.

**Comment by dhomeier** [comment]
You mean if downstream wants to test agains newest Astropy and NumPy 1.18 or 1.19? That's a somewhat awkward combination; OTOH it is perhaps not that uncommon to test legacy versions of only some packages, leaving all others at their defaults. Still I think it's not too much asked that they pin to LTS Astropy for those test envs.

**Comment by pllim** [comment]
Okaaaay... But we should do this earlier in the cycle next time. Hopefully some sort of automation would help.

**Comment by mhvk** [comment]
Agree that we should do these things post-release.

File: astropy/utils/masked/tests/test_function_helpers.py [hunk 1]
```diff
@@ -1341,18 +1338,7 @@ def test_nanpercentile(self):
...
-    untested_functions |= financial_functions
-
-if NUMPY_LT_1_23:
-    deprecated_functions = {
-        # Deprecated, removed in numpy 1.23
-        np.asscalar, np.alen,
-    }
-else:
-    deprecated_functions = set()
+deprecated_functions = set()
```
Comment (eerovaher) [inline astropy/utils/masked/tests/test_function_helpers.py:1341]: Is this intentional?

Comment (mhvk) [reply astropy/utils/masked/tests/test_function_helpers.py]: Oops, no!! I guess without testing all numpy versions one doesn't see those kinds of things... Now pushed a corrected version.

**Comment by astrofrog** [comment]
@pllim - well to give advance warning I think we can require Python 3.9+ and Numpy 1.21+ in ``main`` once 5.2 is released! 😄

**Comment by dhomeier** [comment]
> I don't think it is very likely that people would test against -dev with an old version of numpy. Also, things should just work with newer numpy, also for their pacakges!

Well it would hit them with -stable once 5.2 is released, right? Still don't think that would need to be tested together with numpy-legacy.

**Comment by pllim** [comment]
Can you please rebase? We just merged a few things to fix the CI. Thanks!

**Comment by pllim** [comment]
Also I think this needs a change log under "others" to mentioned the minversion bumps. I think users need to know that.

**Comment by mhvk** [comment]
OK, rebased. The coverage failure is a false positive. I'm quite happy to just put this in -- the one small remaining issue is whether to raise the minimum version of scipy to 1.6 so we do not need the modeling xfail for one test.

**Comment by eerovaher** [comment]
We shouldn't forget to update the known issues page, see e.g. https://github.com/astropy/astropy/blob/e32652cf0e1d9dcfda9ec9ae1193e7ea3ee40788/docs/known_issues.rst#L80-L89

**Comment by pllim** [comment]
@mhvk , do you also wanna clean up the known issue page as @eerovaher suggested above in this same PR? Thanks!

**Review by WilliamJamieson — APPROVED** [review]

**Comment by mhvk** [comment]
> do you also wanna clean up the known issue page as @eerovaher suggested above in this same PR?

No, I'd rather have it be a separate issue (I did have a quick look and don't think there actually are any changes, but don't have the time to check properly)

**Review by pllim — APPROVED** [review]
Changes LGTM. Thanks!

I agree known issues can be a follow-up PR. Let's get this in sooner than later.

**Pull request merged**